spyre-rag/src/ingest/toc.py

```diff
@@ -0,0 +1,157 @@
+            Mapping from an object id to the number of the page that contains
+            that object.
+    """
+
+
+    def __init__(self, document: PDFDocument):
+        self.document = document
+        # obj_id -> page_number
+        self.objid_to_pagenum: dict[int, int] = {
+            page.pageid: page_num
+            for page_num, page in enumerate(PDFPage.create_pages(document), 1)
+        }
+
+    @classmethod
+    def get_ref_type(cls, ref: Any) -> PDFRefType:
+        """Get the type of a PDF reference."""
+        if isinstance(ref, PDFObjRef):
+            return PDFRefType.PDF_OBJ_REF
+        elif isinstance(ref, dict) and "D" in ref:
+            return PDFRefType.DICTIONARY
+        elif isinstance(ref, list) and any(isinstance(e, PDFObjRef) for e in ref):
+            return PDFRefType.LIST
+        elif isinstance(ref, bytes):
+            return PDFRefType.NAMED_REF
+        else:
+            return PDFRefType.UNK
+
+
+    @classmethod
+    def is_ref_page(cls, ref: Any) -> bool:
+        """Check whether a reference is of type '/Page'.
+
+
+        Args:
+            ref (:obj:`Any`):
+                The PDF reference.
+
+        Returns:
+            :obj:`bool`: :obj:`True` if the reference references
+            a page, :obj:`False` otherwise.
+        """
+        return isinstance(ref, dict) and "Type" in ref and ref["Type"] is LITERAL_PAGE
+
+    def resolve(self, ref: Any) -> Optional[int]:
+        """Resolve a PDF reference to a page number recursively.
+
+        Args:
+            ref (:obj:`Any`):
+                The PDF reference.
+
+        Returns:
+            :obj:`Optional[int]`: The page number or :obj:`None`
+            if the reference could not be resolved (e.g., remote Go-To
+            Actions or malformed references).
+        """
+        ref_type = self.get_ref_type(ref)
+
+        if ref_type is PDFRefType.PDF_OBJ_REF and self.is_ref_page(ref.resolve()):
+            return self.objid_to_pagenum.get(ref.objid)
+        elif ref_type is PDFRefType.PDF_OBJ_REF:
+            return self.resolve(ref.resolve())
+
+        if ref_type is PDFRefType.DICTIONARY:
+            return self.resolve(ref["D"])
+
+        if ref_type is PDFRefType.LIST:
+            # Get the PDFObjRef in the list (usually first element).
+            return self.resolve(next(filter(lambda e: isinstance(e, PDFObjRef), ref)))
+
+        if ref_type is PDFRefType.NAMED_REF:
+            return self.resolve(self.document.get_dest(ref))
+
+        return None  # PDFRefType.UNK
+
+def get_matching_header_lvl(toc, title, page_no, threshold=80):
+    titleAndLevel = toc.get(page_no, None)
+    if not titleAndLevel:
+        return ""
+    
+    tocHeader = titleAndLevel["title"]
+    score = fuzz.partial_ratio(title.lower(), tocHeader.lower())
+    if score >= threshold:
+        return "#" * int(titleAndLevel["level"])
+    return ""
+
+def get_toc(file):
+    toc = {}
+    with open(file, "rb") as fp:
+        try:
```

Comment:
Here exception handling doesn't cover other possible exceptions. Is that intended?